File: playground/src/examples/primitives/slot.ts
import { Slot } from "@tevm/voltaire";
// Example: Slot basics - Consensus layer slot numbers
// Slots are the fundamental unit of time in Ethereum's proof-of-stake consensus
// Each slot is 12 seconds, and 32 slots form one epoch

// Create slots using various input types
const genesisSlot = Slot(0);
const slot1M = Slot(1000000);
const slotFromBigInt = Slot(7000000n);
const slotFromHex = Slot("0xf4240"); // 1000000
const slotFromString = Slot("5000000");

// Epoch boundaries (32 slots per epoch)
const epochBoundaries = [
	Slot(0), // Epoch 0, slot 0
	Slot(32), // Epoch 1, slot 0
	Slot(64), // Epoch 2, slot 0
	Slot(96), // Epoch 3, slot 0
];
for (const slot of epochBoundaries) {
	const epoch = Slot.toEpoch(slot);
}
const slotsInEpoch = [
	Slot(32), // First slot of epoch 1
	Slot(33), // Second slot of epoch 1
	Slot(40), // Middle of epoch 1
	Slot(63), // Last slot of epoch 1
	Slot(64), // First slot of epoch 2
];

for (const slot of slotsInEpoch) {
	const epoch = Slot.toEpoch(slot);
	const slotInEpoch = Number(slot) % 32;
}
const currentSlot = Slot(7000000n);
const nextSlot = Slot(7000001n);
const prevSlot = Slot(6999999n);
const slotA = Slot(1000000n);
const slotB = Slot(1000000n);
const slotC = Slot(1000001n);
const realisticSlots = [
	{ slot: Slot(7000000n), label: "~Mid 2024" },
	{ slot: Slot(7500000n), label: "~Late 2024" },
	{ slot: Slot(8000000n), label: "~Early 2025" },
];

for (const { slot, label } of realisticSlots) {
	const epoch = Slot.toEpoch(slot);
}

File: packages/voltaire-zig/src/primitives/Slot/Slot.zig
//! Slot - Consensus Layer Slot Number
//!
//! Represents a consensus layer slot number (12 seconds per slot).
//! Slots are the fundamental unit of time in Ethereum's proof-of-stake consensus.
//!
//! Each slot has a duration of 12 seconds, and 32 slots form an epoch.
//!
//! ## Examples
//!
//! ```zig
//! const primitives = @import("primitives");
//! const Slot = primitives.Slot;
//!
//! // Create slot
//! const slot = Slot.from(1000000);
//!
//! // Convert to number
//! const num = Slot.toNumber(slot);
//!
//! // Convert to epoch (slot / 32)
//! const epoch = Slot.toEpoch(slot);
//! ```

const std = @import("std");
const Epoch = @import("../Epoch/Epoch.zig");

/// Slot type - represents a consensus layer slot number
pub const Slot = u64;

/// Create Slot from u64 value
///
/// ## Parameters
/// - `value`: u64 slot number
///
/// ## Returns
/// Slot value
///
/// ## Example
/// ```zig
/// const slot = Slot.from(1000000);
/// ```
pub fn from(value: u64) Slot {
    return value;
}

/// Convert Slot to u64
///
/// ## Parameters
/// - `slot`: Slot value
///
/// ## Returns
/// u64 representation
///
/// ## Example
/// ```zig
/// const slot = Slot.from(1000000);
/// const num = Slot.toNumber(slot);
/// ```
pub fn toNumber(slot: Slot) u64 {
    return slot;
}

/// Check if two Slots are equal
///
/// ## Parameters
/// - `a`: First slot
/// - `b`: Second slot
///
/// ## Returns
/// true if equal
///
/// ## Example
/// ```zig
/// const a = Slot.from(1000000);
/// const b = Slot.from(1000000);
/// const result = Slot.equals(a, b); // true
/// ```
pub fn equals(a: Slot, b: Slot) bool {
    return a == b;
}

/// Convert Slot to its corresponding Epoch
///
/// Each epoch contains 32 slots. This performs integer division: epoch = slot / 32.
///
/// ## Parameters
/// - `slot`: Slot value
///
/// ## Returns
/// Epoch value
///
/// ## Example
/// ```zig
/// const slot = Slot.from(96);
/// const epoch = Slot.toEpoch(slot); // 3 (96 / 32)
/// ```
pub fn toEpoch(slot: Slot) Epoch.Epoch {
    return Epoch.from(slot / 32);
}

// ====================================
// Tests
// ====================================

test "Slot.from creates slot" {
    const slot = from(1000000);
    try std.testing.expectEqual(1000000, slot);
}

test "Slot.toNumber converts to u64" {
    const slot = from(1000000);
    try std.testing.expectEqual(1000000, toNumber(slot));
}

test "Slot.equals returns true for equal slots" {
    const a = from(1000000);
    const b = from(1000000);
    try std.testing.expect(equals(a, b));
}

test "Slot.equals returns false for unequal slots" {
    const a = from(1000000);
    const b = from(1000001);
    try std.testing.expect(!equals(a, b));
}

test "Slot.toEpoch converts slot to epoch" {
    const slot = from(96);
    const epoch = toEpoch(slot);
    try std.testing.expectEqual(3, epoch);
}

test "Slot.toEpoch rounds down for partial epochs" {
    const slot = from(97);
    const epoch = toEpoch(slot);
    try std.testing.expectEqual(3, epoch);
}

test "Slot.toEpoch handles slot 0" {
    const slot = from(0);
    const epoch = toEpoch(slot);
    try std.testing.expectEqual(0, epoch);
}

test "Slot.toEpoch handles first slot of epoch" {
    const slot = from(32);
    const epoch = toEpoch(slot);
    try std.testing.expectEqual(1, epoch);
}

test "Slot.toEpoch handles large slot numbers" {
    const slot = from(1000000);
    const epoch = toEpoch(slot);
    try std.testing.expectEqual(31250, epoch);
}
